Vmd pch rootbus

sdurawa opened the pull request:
VMD PCH rootbus implementation

Commit:
PCI: vmd: Add WA for VMD PCH rootbus support

VMD PCH rootbus primary number is 0x80 and pci_scan_bridge_extend()
cannot assign it as "hard-wired to 0" and marks setup as broken. To
avoid this, PCH bus number has to be the same as PCH primary number.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -34,6 +34,15 @@
+	VMD_RESOURCE_COUNT
+};
+
+enum vmd_rootbus { VMD_ROOTBUS_0 = 0, VMD_ROOTBUS_COUNT };
```
I've never seen something like that before, probably visual formatting.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -505,7 +514,8 @@ static inline void vmd_acpi_end(void) { }
 {
-	u16 bus, max_buses = resource_size(&vmd->resources[0]);
+	u16 bus,
+		max_buses = resource_size(&vmd->resources[VMD_RESOURCE_CFGBAR]);
```
not acceptable:

```suggestion
u16 bus;
u16 max_buses = resource_size(&vmd->resources[VMD_RESOURCE_CFGBAR]);
```

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -553,8 +563,10 @@ static void vmd_domain_reset(struct vmd_dev *vmd)
-	vmd->dev->resource[VMD_MEMBAR1].child = &vmd->resources[1];
-	vmd->dev->resource[VMD_MEMBAR2].child = &vmd->resources[2];
+	vmd->dev->resource[VMD_MEMBAR1].child =
+		&vmd->resources[VMD_RESOURCE_MEMBAR_1];
```
For sure, it is badly formatted. 

```suggestion
struct x *vmd_membar1_res=&vmd->dev->resource[VMD_MEMBAR1];
struct x *vmd_membar2_res =&vmd->dev->resource[VMD_MEMBAR2];

vmd_membar1_res->child =&vmd->dev->resource[VMD_MEMBAR1];
vmd_membar2_res->child =&vmd->dev->resource[VMD_MEMBAR2];
```

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +779,126 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
-static int vmd_enable_domain(struct vmd_dev *vmd, unsigned long features)
+static void vmd_configure_cfgbar(struct vmd_dev *vmd)
+{
+	struct resource *res;
```
```suggestion
struct resource *res = &vmd->dev->resource[VMD_CFGBAR];
```

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +779,126 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+ * @membar_number: type of the MemBAR
+ * @start_offset: 4K aligned offset applied to start of VMD’s MEMBAR MMIO space
+ * @end_offset: 4K aligned offset applied to end of VMD’s MEMBAR MMIO space
+ * @parent: parent resource assigned to resource to be filled in
```
Highlight that it can be NULL.

mtkaczyk replied:
I mean `@parent` lol :)

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +779,126 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+ * to MMIO address assigned by the OS or BIOS.
+ * @vmd: the VMD device
+ * @resource_number: resource buffer number to be filled in
+ * @membar_number: type of the MemBAR
```
you say `type` but the name has `number`

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +779,126 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	snprintf(name, sizeof(name), "VMD MEMBAR%d", membar_number/2);
+
+	if (!parent)
+		parent = res;
```
generally overwriting income parameter is bad practice, maybe better. use inline if?

```
		.parent = parent ? parent : res;
```

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -892,70 +983,39 @@ static int vmd_enable_domain(struct vmd_dev *vmd, unsigned long features)
-	pci_add_resource_offset(&resources, &vmd->resources[2], offset[1]);
+	pci_add_resource(&resources, &vmd->resources[VMD_RESOURCE_CFGBAR]);
+	pci_add_resource_offset(
+		&resources, &vmd->resources[VMD_RESOURCE_MEMBAR_1], offset[0]);
```
looks like bad formatting to me, please look into kernel, how they solve this :)

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -892,70 +983,39 @@ static int vmd_enable_domain(struct vmd_dev *vmd, unsigned long features)
-	vmd->bus = pci_create_root_bus(&vmd->dev->dev, vmd->busn_start,
-				       &vmd_ops, sd, &resources);
-	if (!vmd->bus) {
+	vmd->bus[VMD_ROOTBUS_0] = pci_create_root_bus(
```
bad formatting.

mtkaczyk reviewed (commented):
patch 1 review
for title I suggest to remove reference to Intel CPU code number (not needed here):
```
PCI: vmd: Clean up vmd_enable_domain function
This function is too long and needs to be shortened to make it more readable.

This clean up is a prework for enablement additional VMD bus range. It doesn't change
Functional behavior of vmd_enable_domain().

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

```

typo `beahvior`

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -519,6 +543,11 @@ static void vmd_domain_reset(struct vmd_dev *vmd)
 	char __iomem *base;
 
+	if (vmd_has_pch_rootbus(vmd))
+		max_buses += resource_size(
```
bad formating

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -664,6 +693,19 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+		case 3:
+			if (!(features & VMD_FEAT_HAS_PCH_ROOTBUS)) {
+				pci_err(dev,
+					"VMD Bus Restriction detected type %d, but PCH Rootbus is not supported, aborting.\n",
```
You can move it up to previous line, lgtm anyway

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -791,6 +835,31 @@ static void vmd_configure_cfgbar(struct vmd_dev *vmd)
+		 * PCI_REG_BUSRANGE0 register.
+		 */
+		vmd->resources[VMD_RESOURCE_CFGBAR].start =
+			VMD_ROOTBUS_RANGE_START(ioc_bus_range);
```
I'm not sure if this is correct in kernel coding style :(

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -839,10 +909,44 @@ static void vmd_configure_membar(struct vmd_dev *vmd,
+				     &vmd->resources[VMD_RESOURCE_MEMBAR_1]);
+		vmd_configure_membar(vmd, VMD_RESOURCE_PCH_MEMBAR_2,
+				     VMD_MEMBAR2,
+				     membar2_offset + pch_membar2_offset, 0,
```
consider shorter naming for variables.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -400,8 +400,21 @@ static inline u8 vmd_has_pch_rootbus(struct vmd_dev *vmd)
+	 * which is stored in vmd->busn_start[VMD_ROOTBUS_1].
+	 */
+	bus_number = (bus->number == VMD_PRIMARY_PCH_BUS) ?
+			     vmd->busn_start[VMD_ROOTBUS_1] :
```
formating

Commit:
PCI: vmd: Clean up vmd_enable_domain function

This function is too long and needs to be shortened to make it more readable.
Certain VMD devices like Intel Arrow Lake has two bus ranges, current
implementation supports only one range. This clean up is a prework for
enablement additional VMD bus range. It doesn't change functional
behavior of vmd_enable_domain().

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Add VMD PCH rootbus support

Starting from Intel Arrow Lake VMD enhacement introduces separate
rotbus for PCH. It means that all 3 MMIO BARs exposed by VMD are
shared now between CPU IOC and PCH. This patch adds PCH bus
enumeration and MMIO management for devices with VMD enhancement
support.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +832,235 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	pci_add_resource_offset(&resources_pch,
+				&vmd->resources[VMD_RES_PCH_MEMBAR_2], offset[1]);
+
+	vmd->bus[VMD_ROOTBUS_1] = pci_create_root_bus(
```
this formatting is bad.
